library/kernel_settings.py

```diff
@@ -0,0 +1,853 @@
+        )
+        orig_gc_load_config_defaults = (
+            tuned.utils.global_config.GlobalConfig.load_config.__func__.__defaults__
+        )
+        if orig_gc_init_defaults:
+            tuned.utils.global_config.GlobalConfig.__init__.__func__.__defaults__ = (
+                tuned.consts.GLOBAL_CONFIG_FILE,
+            )
+        if orig_gc_load_config_defaults:
+            tuned.utils.global_config.GlobalConfig.load_config.__func__.__defaults__ = (
+                tuned.consts.GLOBAL_CONFIG_FILE,
+            )
+    else:
+        orig_gc_init_defaults = (
+            tuned.utils.global_config.GlobalConfig.__init__.__defaults__
+        )
+        orig_gc_load_config_defaults = (
+            tuned.utils.global_config.GlobalConfig.load_config.__defaults__
+        )
+        if orig_gc_init_defaults:
+            tuned.utils.global_config.GlobalConfig.__init__.__defaults__ = (
+                tuned.consts.GLOBAL_CONFIG_FILE,
+            )
+        if orig_gc_load_config_defaults:
+            tuned.utils.global_config.GlobalConfig.load_config.__defaults__ = (
+                tuned.consts.GLOBAL_CONFIG_FILE,
+            )
+    # this call fails on ubuntu and containers, so mock it for testing
+    import pyudev.monitor
+
+    orig_set_receive_buffer_size = pyudev.monitor.Monitor.set_receive_buffer_size
+    pyudev.monitor.Monitor.set_receive_buffer_size = lambda self, size: None
+
+    def test_cleanup():
+        import os
+        import shutil
+
+        if "TEST_ROOT_DIR" not in os.environ:
+            shutil.rmtree(test_root_dir)
+        for cnst, val in orig_consts.items():
+            tuned.consts.__dict__[cnst] = val
+        for idx, dname in enumerate(orig_load_dirs):
+            tuned.consts.LOAD_DIRECTORIES[idx] = dname
+        for idx, dname in enumerate(orig_rec_dirs):
+            tuned.consts.RECOMMEND_DIRECTORIES[idx] = dname
+        if has_func:
+            tuned.utils.global_config.GlobalConfig.__init__.__func__.__defaults__ = (
+                orig_gc_init_defaults
+            )
+            tuned.utils.global_config.GlobalConfig.load_config.__func__.__defaults__ = (
+                orig_gc_load_config_defaults
+            )
+        else:
+            tuned.utils.global_config.GlobalConfig.__init__.__defaults__ = (
+                orig_gc_init_defaults
+            )
+            tuned.utils.global_config.GlobalConfig.load_config.__defaults__ = (
+                orig_gc_load_config_defaults
+            )
+        pyudev.monitor.Monitor.set_receive_buffer_size = orig_set_receive_buffer_size
+
+    if "TEST_ROOT_DIR" not in os.environ:
+        atexit.register(test_cleanup)
+    return test_cleanup
+
+
+def get_tuned_config():
+    """get the tuned config and set our parameters in it"""
+    tuned_config = tuned.utils.global_config.GlobalConfig()
+    tuned_config.set("daemon", 0)
+    tuned_config.set("reapply_sysctl", 0)
```

Comment:
> Posting some minor comments...
> 
> One thing that just occured to me: instead of using the internal tuned API, couldn't you just do something along the lines of the following?
> 
> ```
> < update the kernel_settings profile according to the module parameters >
> 
> if ! grep kernel_settings /etc/tuned/active_profile; then
>     tuned-adm profile kernel_settings $(tuned-adm active)
> else
>     systemctl restart tuned
> fi
> ```
> 
> (Even the RHEL-7 Tuned seems to support applying multiple profiles.) Sorry if this has been raised before.

Not sure what you mean.  Do you mean to implement the kernel_settings module without the python wrapper around the tuned code e.g. as a series of ansible tasks that invoke modules such as `ini_file`, and shell commands?  If so, I think that is something we should do in the future, if possible.

Or do you mean specifically the bash code excerpt above about managing the active profile and restarting tuned, and that we could do that instead of https://github.com/linux-system-roles/kernel_settings/pull/1/files#diff-ca666a588d70612a44b9f4ef999c9525R495 ?  If so, I'm not sure what that would gain us.  Also note that in ansible, the restart of tuned is conditional.  If the module detects that something has changed, it will return a value of `"changed":true` in the result: https://github.com/linux-system-roles/kernel_settings/pull/1/files#diff-ca666a588d70612a44b9f4ef999c9525R833
The task will notify the handler if the module returns `"changed":true` https://github.com/linux-system-roles/kernel_settings/pull/1/files#diff-2444ad0870f91f17ca6c2a5e96b26823R38

Reply:
Actually, never mind. What I was really wondering is how is the "profile pre-load" feature (https://bugzilla.redhat.com/show_bug.cgi?id=1798183) is going to help you, given that there is already the possibility to apply multiple profiles. And I think the answer to that is that you want the `kernel_settings` profile to remain applied even if the user manually runs `tuned-adm profile <some other profile>`. Right?

Reply:
> And I think the answer to that is that you want the kernel_settings profile to remain applied even if the user manually runs tuned-adm profile <some other profile>. Right?

Right.  So for now, we are relying on the user not to remove the kernel_settings profile, or otherwise relying on the user to not change tuned config in such a way that kernel_settings are not applied.  Note that this is not a new problem, and not a problem with the kernel_settings role specifically (e.g. the network role - the user can go in and manually change network settings, outside of the role, that will cause the settings applied by the network role do not function correctly) - but we don't want it to be too easy for the user to alter the kernel_settings using `tuned-adm` or editing tuned profiles.